File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
        ListNode* reverseList(ListNode* head) {
            stack<int> s;
            ListNode *temp = new ListNode() , *newHead = new ListNode(); 
            temp = head;
            while(temp != NULL){
                s.push(temp->val); 
                temp = temp-> next;
            }
            temp = newHead;
            while(!s.empty()){
                temp->next = new ListNode(s.top());
                s.pop();
                temp = temp->next;
            }
            return newHead->next;
        }
};

File: Data Structures & Algorithms/middle-of-the-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        ListNode dummy = new ListNode(-1,head);
        ListNode first = dummy;
        ListNode second = head;
        int n = 0;
        while(first.next != null){
                first = first.next;
                n+=1;
        } 
        System.out.print(n);
        n/=2;
        while(n > 0){
            second = second.next;
            n-=1;
        }
        return second;
    }
}